### PR TITLE
Add e2e testing to spdy proxy support using netexec

### PR DESCRIPTION
**Pull request by ashcrow**
Verified using ginkgo against a local cluster as originally discussed. This supersedes #3.
## Changes
- rebased on kubernetes/master (hence the non automatic merge)
- https://github.com/ncdc/kubernetes/commit/2b4dcdf08b31c46992a4411841d6e854911ad292: url.URL.Host now canonical before use in spdy roundtripper 
- https://github.com/ncdc/kubernetes/commit/c7b1a48e7c80d1c74d6953bf6c0e28bafb939e91: Fixes netexec Makefiles TAG variable: 
- https://github.com/ncdc/kubernetes/commit/9e4ef014ca491693f7a8b845fe323e75516fe335
  - Adds upload functionality
  - Adds EXPOSE to the Dockerfile
- https://github.com/ncdc/kubernetes/commit/704e1a08566524b008ad6790023d6399b645964b: Adds a pod.yaml for netexec
- https://github.com/ncdc/kubernetes/commit/017d11ef17ff83539ac8dbe72c65141f9f476f79: Adds more logging for shell handler: 
- https://github.com/ncdc/kubernetes/commit/f53d5cab6805d35e643b0f7d939cd744919ca48b
  - pod file for goproxy added 
  - EXPOSE added to Dockerfile
- https://github.com/ncdc/kubernetes/commit/b79bc657c89cce046460b402b5faf128b0377ef9: e2e updates in `test/e2e/kubectl.go` to enable testing: 
## Notes

e2e requires two new environment variables.
- `KUBECTLBIN`: The full path to a _static_ `kubectl` binary
- `APISERVERADDR`: Full URL to the API server (IE: http://10.10.10.10:8080). This is fed in to the exec-proxy-test pod as its `kubectl --server` option.

### CODE REVIEW

File: contrib/for-tests/netexec/netexec.go [hunk 1]
```diff
@@ -191,11 +193,53 @@ func dialUDP(request string, remoteAddress *net.UDPAddr) (string, error) {
 
 func shellHandler(w http.ResponseWriter, r *http.Request) {
 	log.Println(r.FormValue("shellCommand"))
-	output, err := exec.Command(shellPath, "-c", r.FormValue("shellCommand")).CombinedOutput()
-	assertNoError(err)
```
Comment (ncdc) [inline contrib/for-tests/netexec/netexec.go:195]: Why did you remove the assert and replace it with a log?

File: contrib/for-tests/netexec/netexec.go [hunk 2]
```diff
@@ -191,11 +193,53 @@ func dialUDP(request string, remoteAddress *net.UDPAddr) (string, error) {
...
+	if _, err = io.Copy(f, file); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("Unable to write file."))
+		log.Printf("Unable to write file: %s", err)
+		return
+	}
+
+	UploadFile := f.Name()
+	err = os.Chmod(UploadFile, 0700)
+	if err != nil {
```
Comment (ncdc) [inline contrib/for-tests/netexec/netexec.go]: if err := os.Chmod(...); err != nil {

File: test/e2e/kubectl.go [hunk 3]
```diff
@@ -188,6 +195,114 @@ var _ = Describe("Kubectl client", func() {
...
+			if err != nil {
+				Failf("Unable to get the goproxy pod. Error: %s", err)
+			}
+			proxyAddr := fmt.Sprintf("http://%s:8080", goproxyPod.Status.PodIP)
+			// get the exec address
+			netexecPod, err := c.Pods(ns).Get(netexecContainer)
+			netexecAddr := fmt.Sprintf("http://%s:8080", netexecPod.Status.PodIP)
+
+			// upload kubectl to remote exec server
+			postBodyBuffer := &bytes.Buffer{}
```
Comment (ncdc) [inline test/e2e/kubectl.go]: We may want to consider using a pipe to stream the upload instead of copying the entire kubectl binary into memory prior to uploading. See https://github.com/gebi/go-fileupload-example/blob/master/main.go for an example.

File: test/e2e/kubectl.go [hunk 4]
```diff
@@ -188,6 +195,114 @@ var _ = Describe("Kubectl client", func() {
...
+			_, err = io.Copy(fileWriter, fh)
+			if err != nil {
+				Failf("Unable to to copy kubectl at %s into the file writer. Error: %s", kubectlBinPath, err)
+			}
+			// Nothing more should be written to this instance of the postBodyWriter
+			postBodyWriter.Close()
+
+			resp, err := http.Post(fmt.Sprintf("%s/upload", netexecAddr), postBodyWriter.FormDataContentType(), postBodyBuffer)
+
+			uploadBinaryName, err := ioutil.ReadAll(resp.Body)
```
Comment (ncdc) [inline test/e2e/kubectl.go]: Check `err` first. If it's non-nil, `Failf`. Next check `resp.StatusCode`. Finally, read the body.

File: test/e2e/kubectl.go [hunk 5]
```diff
@@ -188,6 +195,114 @@ var _ = Describe("Kubectl client", func() {
 			if e, a := "hi", execOutput; e != a {
 				Failf("Unexpected kubectl exec output. Wanted %q, got %q", e, a)
 			}
+
+			By("executing a command in the container through an HTTP proxy")
```
Comment (ncdc) [inline test/e2e/kubectl.go]: I'm wondering if it would make more sense for this to be an `It()` sibling to `It("should support exec"`

File: test/e2e/kubectl.go [hunk 6]
```diff
@@ -188,6 +195,114 @@ var _ = Describe("Kubectl client", func() {
...
+			}
+
+			// Copy kubectl binary into the file writer
+			_, err = io.Copy(fileWriter, fh)
+			if err != nil {
+				Failf("Unable to to copy kubectl at %s into the file writer. Error: %s", kubectlBinPath, err)
+			}
+			// Nothing more should be written to this instance of the postBodyWriter
+			postBodyWriter.Close()
+
```
Comment (ncdc) [inline test/e2e/kubectl.go]: Add a `By` indicating you're uploading

File: test/e2e/kubectl.go [hunk 7]
```diff
@@ -188,6 +195,114 @@ var _ = Describe("Kubectl client", func() {
...
+			if resp.StatusCode != http.StatusCreated {
+				Failf("Unable to upload kubectl binary to remote exec server. Expected status: 201, got %d", resp.StatusCode)
+			}
+			if err != nil {
+				Failf("Unable to upload kubectl binary to the remote exec server due to error: %s", err)
+			}
+			for _, proxyVar := range []string{"http_proxy", "HTTP_PROXY", "https_proxy", "HTTPS_PROXY"} {
+				shellCommand := fmt.Sprintf("%s=%s %s --server=%s --namespace=%s exec nginx echo running in container", proxyVar, proxyAddr, uploadBinaryName, apiserverAddr, ns)
+				// Execute kubectl on remote exec server.
+				resp, err = http.PostForm(fmt.Sprintf("%s/shell", netexecAddr), url.Values{"shellCommand": {shellCommand}})
```
Comment (ncdc) [inline test/e2e/kubectl.go]: Add a `By` indicating you're running exec via proxy and include the env var name

File: test/e2e/kubectl.go [hunk 8]
```diff
@@ -188,6 +195,114 @@ var _ = Describe("Kubectl client", func() {
...
+				if resp.StatusCode != http.StatusOK {
+					errOut, _ := ioutil.ReadAll(resp.Body)
+					Failf("Unable to execute kubectl binary on remote exec server. Expected status: %d, got %d. Info: %s", http.StatusCreated, resp.StatusCode, errOut)
+				}
+
+				proxyExecOutput, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					Failf("Unable to read the result of the kubectl execution on the remote exec server due to error: %s", err)
+				}
+				resp.Body.Close()
```
Comment (ncdc) [inline test/e2e/kubectl.go]: I'd `defer` this and move it just before the check on `resp.StatusCode`

File: test/e2e/kubectl.go [hunk 9]
```diff
@@ -188,6 +195,114 @@ var _ = Describe("Kubectl client", func() {
...
+				// Verify we got the normal output captured by the exec server
+				if string(proxyExecOutput) != expectedExecOutput {
+					Failf("Unexpected kubectl exec output. Wanted '%s', got '%s'", expectedExecOutput, proxyExecOutput)
+				}
+
+				// Verify the proxy server logs saw the connection
+				expectedProxyLog := fmt.Sprintf("Sending request GET %s/api", apiserverAddr)
+				rawProxyLog := strings.Split(runKubectl("log", "goproxy", fmt.Sprintf("--namespace=%v", ns)), "\n")
+				sort.Sort(sort.Reverse(sort.StringSlice(rawProxyLog)))
+				proxyLog := rawProxyLog[2]
```
Comment (ncdc) [inline test/e2e/kubectl.go]: Could you just do `strings.Contains` against `rawProxyLog`? I'm concerned about relying on a specific line number in the log.

File: contrib/for-tests/netexec/netexec.go (hunk 1 above)
Comment (ashcrow) [inline contrib/for-tests/netexec/netexec.go:195]: To control the output of the logging. It's not a must have by any means but I found it a bit easier to read in the logs.

File: contrib/for-tests/netexec/netexec.go (hunk 2 above)
Comment (ashcrow) [inline contrib/for-tests/netexec/netexec.go]: Will update.

File: test/e2e/kubectl.go (hunk 9 above)
Comment (ashcrow) [inline test/e2e/kubectl.go]: I originally went that route. The problem was that the logs continue to build up meaning that if the first `proxyVar` is successful then `strings.Contains` will always say the following ones were as well. Another option would be to redeploy the `goproxy` pod for each `proxyVar` but I fear it would add significant time to the test.

File: test/e2e/kubectl.go (hunk 8 above)
Comment (ashcrow) [inline test/e2e/kubectl.go]: :+1:

File: test/e2e/kubectl.go (hunk 9 above)
Comment (ncdc) [inline test/e2e/kubectl.go]: I was assuming a redeploy for each test. It's fine to leave it like this for now, but let's be prepared to change to a redeploy once we get more eyes on this.

File: test/e2e/kubectl.go (hunk 6 above)
Comment (ashcrow) [inline test/e2e/kubectl.go]: Will add.

File: test/e2e/kubectl.go (hunk 7 above)
Comment (ashcrow) [inline test/e2e/kubectl.go]: Will add.

File: test/e2e/kubectl.go (hunk 9 above)
Comment (ashcrow) [inline test/e2e/kubectl.go]: I'll go ahead and do it now since it sounds like it will be needed anyway. No time like the present :smile:

**Comment by ashcrow** [comment]
This should now have the feedback incorporated. For the moment I've commented out HTTPS/https testing. Will continue looking Monday (and see if Clayton has an idea as well).

File: contrib/for-tests/netexec/netexec.go (hunk 1 above)
Comment (ncdc) [inline contrib/for-tests/netexec/netexec.go:195]: `assertNoError` calls `log.Fatal` if err is non-nil, which will terminate the executable. Do think that's appropriate, or is it ok to keep things running if there is an error?

Comment (ashcrow) [inline contrib/for-tests/netexec/netexec.go:195]: I'd like it to keep running so that the client gets a response but, then again, it doesn't break anything to not respond back to the client for a test like this.

Comment (ncdc) [inline contrib/for-tests/netexec/netexec.go:195]: I think keeping it running makes sense.
